File: 프로그래머스/2/154538. 숫자 변환하기/숫자 변환하기.py
from collections import deque

def solution(x, y, n):
    answer = []
    
    visit = set()
    visit.add(x)
    
    Q = deque()
    Q.append((x,0)) # value 0 
    
    while len(Q) > 0:
        curValue, curDepth = Q.popleft()
        
        # print(curValue, curDepth)
        
        if y == curValue:
            return curDepth
        
        
        # 1. n 더하기
        nxtValue = curValue + n
        
        
        if nxtValue <= y and nxtValue not in visit:
            visit.add(nxtValue)
            Q.append((nxtValue, curDepth+1))
            
        # 2. 2 곱하기
        nxtValue = curValue * 2
        
        if nxtValue <= y and nxtValue not in visit:
            visit.add(nxtValue)
            Q.append((nxtValue, curDepth+1))
            
        # 3. 3 곱하기
        nxtValue = curValue * 3
        
        if nxtValue <= y and nxtValue not in visit:
            visit.add(nxtValue)
            Q.append((nxtValue, curDepth+1))
    
    
    return -1

File: 프로그래머스/2/154538. 숫자 변환하기/숫자 변환하기.java
import java.util.*;

class Solution {
    public int solution(int x, int y, int n) {
        Set<Integer> visit = new HashSet<>();
        Deque<QEl> Q = new ArrayDeque<>();
        
        Q.addFirst(new QEl(x, 0));
        visit.add(x);
        
        while(Q.size()  > 0){
            QEl curEl = Q.pollLast();
            // 0. 종료 조건
            if (curEl.num == y){
                return curEl.time;
            }
            
            // 1. n을 더함
            int nxtValue1 = curEl.num + n;
            if (nxtValue1 <= y && !visit.contains(nxtValue1)){
                Q.addFirst(new QEl(nxtValue1, curEl.time+1));
                visit.add(nxtValue1);
            }
            
            // 2. 2를 곱함
            int nxtValue2 = curEl.num * 2;
            if (nxtValue2 <= y && !visit.contains(nxtValue2)){
                Q.addFirst(new QEl(nxtValue2, curEl.time+1));
                visit.add(nxtValue2);
            }
            
            // 3. 3을 곱함
            int nxtValue3 = curEl.num * 3;
            if (nxtValue3 <= y && !visit.contains(nxtValue3)){
                Q.addFirst(new QEl(nxtValue3, curEl.time+1));
                visit.add(nxtValue3);
            }
        }
        
        return -1;
    }
    
    private class QEl{
        public int num; // 숫자
        public int time; // 횟수
        
        public QEl(int num, int time){
            this.num = num;
            this.time = time;
        }
    }
    
    
}